### PR TITLE
Groupby tests restructure

### LINKED ISSUE #14813: TST: reorg groupby tests
see suggestions here: https://github.com/pandas-dev/pandas/pull/12607#issuecomment-265310458

we have lots and lots of groupby tests and so getting pretty unweidly

so should reorg to a sub-dir setup like we have for other things

eg.

pandas/tests/groupby/.....

### PR DESCRIPTION
**Pull request by aileronajay**
closes #14813

### CODE REVIEW

**Comment by aileronajay** [comment]
@jreback i have created this pr for the groupby tests restructure, at this point i have only moved the group by filter tests to a new file. Wanted to confirm if this is how it needs to be done also for the review comments.

**Comment by jreback** [comment]
you can name test_groupby_filters -> test_filters.py

yes this is the idea

**Comment by jreback** [comment]
@aileronajay can you update a bit (IOW, split maybe a little more). ping on green and can merge right  away to prevent merge conflicts.

**Comment by aileronajay** [comment]
@jreback i do intend to pick this up from next week

**Comment by aileronajay** [comment]
@jreback i have further split the tests into aggregate and categorical tests. So now we have filters, aggregate and categorical tests in separate files. There is also the original file with the tests that were not moved to one of the new files

**Comment by jreback** [comment]
looks like you have ``test_groupby_filters`` and ``test_filters``?

**Comment by aileronajay** [comment]
@jreback it was a duplicate, i have removed it now

File: pandas/tests/groupby/test_aggregate.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+                   'foo', 'foo', 'foo'],
+             'B': ['one', 'one', 'one', 'two', 'one', 'one', 'one', 'two',
+                   'two', 'two', 'one'],
+             'C': ['dull', 'dull', 'shiny', 'dull', 'dull', 'shiny', 'shiny',
+                   'dull', 'shiny', 'shiny', 'shiny'],
+             'D': np.random.randn(11),
+             'E': np.random.randn(11),
+             'F': np.random.randn(11)})
+
+    def test_agg_api(self):
```
Comment (jreback) [inline pandas/tests/groupby/test_aggregate.py]: can you remove the _agg name from these tests (as its now redundant)

File: pandas/tests/groupby/test_categorical.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        def get_stats(group):
+            return {'min': group.min(),
+                    'max': group.max(),
+                    'count': group.count(),
+                    'mean': group.mean()}
+
+        result = self.df.groupby(cats).D.apply(get_stats)
+        self.assertEqual(result.index.names[0], 'C')
+
+    def test_apply_categorical_data(self):
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: same here on the naming (of course if it adds value when reading the test name, then leave it where it does).

File: pandas/tests/groupby/test_categorical.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+        # GH 10460
+        expc = Categorical.from_codes(
+            np.arange(4).repeat(8), levels, ordered=True)
+        exp = CategoricalIndex(expc)
+        self.assert_index_equal(desc_result.index.get_level_values(0), exp)
+        exp = Index(['count', 'mean', 'std', 'min', '25%', '50%',
+                     '75%', 'max'] * 4)
+        self.assert_index_equal(desc_result.index.get_level_values(1), exp)
+
+    def test_groupby_categorical_index(self):
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: like this one is fine

File: pandas/tests/groupby/test_categorical.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+        # GH 11558
+        cats = pd.CategoricalIndex(['qux', 'foo', 'baz', 'bar'],
+                                   categories=['foo', 'bar', 'baz', 'qux'],
+                                   ordered=True)
+        df = DataFrame(np.random.randn(20, 4), columns=cats)
+        result = df.groupby([1, 2, 3, 4] * 5).describe()
+
+        tm.assert_index_equal(result.columns, cats)
+        tm.assert_categorical_equal(result.columns.values, cats.values)
+
```
Comment (jreback) [inline pandas/tests/groupby/test_categorical.py]: so for the categorical ones, they are mostly ok

File: pandas/tests/groupby/test_filters.py [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+
+    def test_filter_mixed_df(self):
+        df = pd.DataFrame({'A': [1, 12, 12, 1], 'B': 'a b c d'.split()})
+        grouper = df['A'].apply(lambda x: x % 2)
+        grouped = df.groupby(grouper)
+        expected = pd.DataFrame({'A': [12, 12], 'B': ['b', 'c']}, index=[1, 2])
+        assert_frame_equal(
+            grouped.filter(lambda x: x['A'].sum() > 10), expected)
+
+    def test_filter_out_all_groups(self):
```
Comment (jreback) [inline pandas/tests/groupby/test_filters.py]: i would remove _filter as appropriate

**Comment by aileronajay** [comment]
@jreback i had thought about removing the agg (and other identifiers) but then i wonder if leaving them in place is detrimental?

**Comment by jreback** [comment]
@aileronajay so on second thought, let's do that in another PR. then the changes will be obvious. So just leave the split files here. ping on green.

**Comment by aileronajay** [comment]
@jreback all tests are passing

**Comment by jreback** [comment]
thanks @aileronajay 

ok, couple of followups.

originally had 302 tests, now have 305, so a couple duped (not a big deal but if can fix would be great).

move apply, transform, agg from test_groupby -> test_agg (IOW anything having to do with these)

**Comment by aileronajay** [comment]
@jreback i was thinking of moving all cython tests into another file, but the cython tests individually belong to a different category each

**Pull request closed without merging**